### ISSUE TITLE
Invalid model for NRA instances (arith.solver = 6)

### ISSUE THREAD

**Issue by rainoftime** [Z3Prover/z3#3645]
Hi, for the following formula,
~~~~
(set-logic NRA)
(set-option :model_validate true)
(set-option :smt.arith.solver 6)
(declare-const r0 Real)
(declare-const r1 Real)
(push 1)
(assert (< 0.0 r1))
(assert (> 0.0 (- (* r1 r0) r0)))
(check-sat)
~~~~
z3 (commit d8e00bc02e9891) gives an invalid model

**Comment by rainoftime** [comment]
Here is another test case 
~~~~
(set-logic NRA)
(declare-fun _substvar_45_ () Real)
(set-option :model_validate true)
(set-option :combined_solver.ignore_solver1 true)
(set-option :smt.arith.solver 6)
(declare-const r7 Real)
(assert (> (/ r7 r7) _substvar_45_ 0.0))
(check-sat)
~~~~

**Comment by rainoftime** [comment]
Case 3
~~~~
(set-logic QF_UFNRA)
(declare-fun _substvar_28_ () Bool)
(set-option :model_validate true)
(set-option :smt.arith.solver 6)
(declare-const r3 Real)
(declare-const r11 Real)
(push 1)
(assert (>= 73829727737.0 r3 0.82 (/ r11 r11)))
(assert _substvar_28_)
(check-sat)
~~~~

**Comment by NikolajBjorner** [comment]
@levnach - the first example is simple

-tr:arith shows

[(j6 = 2 = (j4 =1)*(j5 =3))
[6]	 [(0, 1),oo]	x = (2, 0) 
root=j6
]
vars:(j4 =1)*(j5 =3)
[4]	 [(0, 1),oo]	x = (1, 0) 
root=j4
[5]	 [(0, 1),oo]	x = (3, 0) base
root=j5

same rvars, and m.rsign = 0 of course
irr: v0 j0 = 1, int := 1
irr: v1 j1 = 1 := 1.0
irr: v2 j2 = 0, int := 0
v3 j3 = 0 := 0.0
v4 j4 = 1 := r1
v5 j6 = 2 := (* r1 r0)
v6 j5 = 3 := r0
v7 j2147483648 = 1 := (+ r0 (* -1.0 (* r1 r0)))

clearly the monomial for j6 is not true (2 is not 1*3).

**Comment by NikolajBjorner** [comment]
The code in 

         void core::patch_monomial_with_real_var(lpvar j)

removes j from the m_to_refine set even if mon_val(m) != var_val(m) for 
the monomial associated with j.

**Comment by levnach** [comment]
Fixed in 0dc5bad6e4300113e1d3e7f393762cb24d72ad29
